### PR TITLE
[Android] Port getifaddrs implementation from Xamarin.Android

### LINKED ISSUE #6973: [Android] NetworkInterface.GetAllNetworkInterfaces() returns strange values on Android 11 + IPv6
---
**Issue moved from dotnet/maui#6649**
- Please respond to @MartinRothschink.

---

_From @MartinRothschink on Friday, April 29, 2022 7:05:05 AM_

### Description

All tests have been done on a Xaomi Pad 5 (physical device, Android 11).

If target sdk is set to < 30, the values returned by NetworkInterface.GetAllNetworkInterfaces() seem valid:
ipv6-28

If target sdk is set to 31, the values returned by NetworkInterface.GetAllNetworkInterfaces() are outside of the enum ranges and name/description are empty. This screen shot shows the situation if IPv6 is not enabled in the router:
ipv4-31

If IPv6 is enabled in the router, NetworkInterfaceType and OperationalStatus values are even worse.
ipv6-31

### Steps to Reproduce

1. Create a new Maui app (VS 2022 17.2 Preview 5)
2. Replace MainPage with the files below

MainPage.xaml.cs

```
namespace MauiApp1;

using System.Net.NetworkInformation;

public partial class MainPage : ContentPage
{
    public MainPage()
    {
        InitializeComponent();
        LoadNics();
        BindingContext = this;
    }

    public Nic[] Nics { get; set; }

    private void LoadNics()
    {
        var nics = new List<Nic>();
        var ifs = NetworkInterface.GetAllNetworkInterfaces();
        foreach (var i in ifs)
        {
            var n = new Nic
            {
                Name = i.Name,
                Description = i.Description,
                Status = i.OperationalStatus,
                Type = i.NetworkInterfaceType
            };

            if (i.Supports(NetworkInterfaceComponent.IPv4))
                n.Supports += "IPv4 ";
            if (i.Supports(NetworkInterfaceComponent.IPv6))
                n.Supports += "IPv6 ";

            foreach (var a in i.GetIPProperties().UnicastAddresses)
            {
                if (a.Address.AddressFamily == System.Net.Sockets.AddressFamily.InterNetworkV6)
                    n.IPv6Addresses += a.Address.ToString() + " ";

                if (a.Address.AddressFamily == System.Net.Sockets.AddressFamily.InterNetwork)
                    n.IPv4Addresses += a.Address.ToString() + " ";
            }

            nics.Add(n);
        }

        Nics = nics.ToArray();
    }
}


public class Nic
{
    public string Name { get; set; }
    public string Description { get; set; }
    public OperationalStatus Status { get; set; }
    public NetworkInterfaceType Type { get; set; }
    public string Supports { get; set;}
    public string IPv4Addresses { get; set; }
    public string IPv6Addresses { get; set; }
}

```

MainPage.xaml

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MauiApp1.MainPage">

    <ListView ItemsSource="{Binding Nics, Mode=OneWay}" SelectionMode="None" RowHeight="{OnPlatform Android=140}">
        <ListView.ItemTemplate>
            <DataTemplate>
                <ViewCell>
                    <Grid ColumnSpacing="8">
                        <Grid.RowDefinitions>
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="Auto" />
                            <RowDefinition Height="Auto" />
                        </Grid.RowDefinitions>
                        <Grid.ColumnDefinitions>
                            <ColumnDefinition Width="Auto" />
                            <ColumnDefinition Width="Auto" />
                            <ColumnDefinition Width="Auto" />
                        </Grid.ColumnDefinitions>
                        <Label Grid.Column="0" Grid.Row="0" Text="Name" FontAttributes="Bold" FontSize="Subtitle"/>
                        <Label Grid.Column="1" Grid.Row="0" Grid.ColumnSpan="2" Text="{Binding Name}" FontSize="Subtitle"/>
                        <Label Grid.Column="1" Grid.Row="1" Text="Description" />
                        <Label Grid.Column="2" Grid.Row="1" Text="{Binding Description}" />
                        <Label Grid.Column="1" Grid.Row="2" Text="Status" />
                        <Label Grid.Column="2" Grid.Row="2" Text="{Binding Status}" />
                        <Label Grid.Column="1" Grid.Row="3" Text="Type" />
                        <Label Grid.Column="2" Grid.Row="3" Text="{Binding Type}" />
                        <Label Grid.Column="1" Grid.Row="4" Text="Supports" />
                        <Label Grid.Column="2" Grid.Row="4" Text="{Binding Supports}" />
                        <Label Grid.Column="1" Grid.Row="5" Text="IPv4Addresses" />
                        <Label Grid.Column="2" Grid.Row="5" Text="{Binding IPv4Addresses}" />
                        <Label Grid.Column="1" Grid.Row="6" Text="IPv6Addresses" />
                        <Label Grid.Column="2" Grid.Row="6" Text="{Binding IPv6Addresses}" />
                    </Grid>
                </ViewCell>
            </DataTemplate>
        </ListView.ItemTemplate>
    </ListView>
</ContentPage>

```

### Version with bug

Release Candidate 2 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 10/11 (target sdk >= 30)

### Did you find any workaround?

Currently I have to ignore the values for OperationalStatus and use the interface with valid UnicastAddresses.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by simonrozsival**
In recent Android versions the data returned by `getifaddrs` is invalid and doesn't allow us to implement `System.Net.NetworkInformation.GetAllNetworkInterfaces()`. It's possible to reimplement `getifaddrs` using Netlink and there's already [an existing implementation in Xamarin.Android](https://github.com/xamarin/xamarin-android/blob/f149c25c7e5936fad73ddf80b588ca67bbacc24a/src/monodroid/jni/xamarin_getifaddrs.cc).

Fixes https://github.com/xamarin/xamarin-android/issues/6973
Ref #62780
Ref #51303

/cc @grendello

### CODE REVIEW

**Review by grendello — APPROVED** [review]

**Comment by simonrozsival** [comment]
The failing tests are unrelated to this PR (on Android x86 they are all `System.Text.Json`, on Linux x64 there's a failing `System.Numerics` test, on Browser there are various failures, mostly HTTP and WebSockets related).

I suggest checking the output of [`System.Net.NetworkInformation.Tests` stdout for android-arm64](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-71943-merge-1c30d3f437e7475ebb/System.Net.NetworkInformation.Functional.Tests/1/testResults.xml?helixlogtype=result) where the network interfaces information of the test device is printed.

**Review by steveisok — APPROVED** [review]

**Comment by steveisok** [comment]
Extra-platforms failures are known and not caused by this change

File: src/mono/cmake/config.h.in [hunk 1]
```diff
@@ -517,9 +517,6 @@
 /* Have getifaddrs */
 #cmakedefine HAVE_GETIFADDRS 1
```
Comment (akoeplinger) [inline src/mono/cmake/config.h.in:518]: @simonrozsival we can also remove this one, it is unused in mono now

Comment (akoeplinger) [reply src/mono/cmake/config.h.in:518]: @simonrozsival that is part of System.Native, not mono. It has its own check: https://github.com/dotnet/runtime/blob/f249dc77c6614e05fe47ac9f5eae987e0ab12036/src/native/libs/configure.cmake#L139-L141

**Pull request merged**